### ISSUE TITLE
Average temp and humidity calculations resulting in saw-tooth rendering in D3

### ISSUE THREAD

**Issue by DanKline** [Azure/connectthedots#233]
I have not been able to pin down the cause of this, but the averages are being plotted as a saw-tooth on the website.

![2016-02-08](https://cloud.githubusercontent.com/assets/3081226/12903438/4397102c-ce8d-11e5-8212-1886bafa6548.png)

**Comment by DanKline** [comment]
I have noticed the errors that are likely causing this behavior, but I am having difficulty pinpointing where in the callbacks the error is created.  The error is being detected at line 670 of D3.js with an invalid path (see attached image)
![image](https://cloud.githubusercontent.com/assets/3081226/13267965/74e7114a-da45-11e5-9198-25dd12f105b2.png)

Any help or guidance would be appreciated.

**Comment by DanKline** [comment]
Here's another capture that demonstrates the issue in the data stream:
Error: Invalid value for <path> attribute d="M0,NaNC0,NaN,0,NaN,0,NaNS0,NaN,0,NaNS61.44723652342241,NaN,61.44723652342241,NaNS61.44723652342241,NaN,61.44723652342241,NaNS92.16959834959353,NaN,92.16959834959353,NaNS92.16959834959353,NaN,92.16959834959353,NaNS92.16959834959353,NaN,92.16959834959353,NaNS92.16959834959353,NaN,92.16959834959353,NaNS122.8894473046845,NaN,122.8894473046845,NaNS153.61683487301593,NaN,153.61683487301593,NaNS153.61683487301593,NaN,153.61683487301593,NaNS184.3341709570267,NaN,184.3341709570267,NaNS215.05653278319784,NaN,215.05653278319784,NaNS215.05653278319784,NaN,215.05653278319784,NaNS245.778894609369,NaN,245.778894609369,NaNS276.5062821777004,NaN,276.5062821777004,NaNS276.5062821777004,NaN,276.5062821777004,NaNS276.5062821777004,NaN,276.5062821777004,NaNS307.24120835927243,NaN,307.24120835927243,NaNS337.94598008788233,NaN,337.94598008788233,NaNS337.94598008788233,NaN,337.94598008788233,NaNS337.94598008788233,NaN,337.94598008788233,NaNS337.94598008788233,NaN,337.94598008788233,NaNS368.6708547851336,NaN,368.6708547851336,NaNS368.6708547851336,NaN,368.6708547851336,NaNS399.3907037402246,NaN,399.3907037402246,NaNS399.3907037402246,NaN,399.3907037402246,NaNS399.3907037402246,NaN,399.3907037402246,NaNS430.11055269531556,NaN,430.11055269531556,NaNS460.8354273925668,NaN,460.8354273925668,NaNS491.557789218738,NaN,491.557789218738,NaNS491.557789218738,NaN,491.557789218738,NaNS522.2751253027487,NaN,522.2751253027487,NaNS522.2751253027487,NaN,522.2751253027487,NaNS553.0100514843207,NaN,553.0100514843207,NaNS553.0100514843207,NaN,553.0100514843207,NaNS583.7223618261711,NaN,583.7223618261711,NaNS614.442210781262,NaN,614.442210781262,NaNS614.442210781262,NaN,614.442210781262,NaNS645.1620597363531,NaN,645.1620597363531,NaNS675.8894473046845,NaN,675.8894473046845,NaNS706.6143220019358,NaN,706.6143220019358,NaNS737.3291452148663,NaN,737.3291452148663,NaNS737.3291452148663,NaN,737.3291452148663,NaNS768.0615585253582,NaN,768.0615585253582,NaNS768.0615585253582,NaN,768.0615585253582,NaNS798.7738688672086,NaN,798.7738688672086,NaN"

**Comment by DanKline** [comment]
It appears that when the vertices are forced to redraw, the d variable on the SVG object are becoming corrupted, possibly by the scale function, and have difficulty recovering.  I'll explore more for a cause and fix.

**Comment by punitganshani** [comment]
@DanKline I'm also facing same issue. The scaleMargin is NaN or Infinity so the value of path "d" has NaN

<img width="770" alt="issue1" src="https://cloud.githubusercontent.com/assets/1032553/13464511/7e50a7b8-e0cc-11e5-81ec-81314aac3358.PNG">

**Comment by DanKline** [comment]
Thanks @punitganshani.  I tried setting a Min and Max value based on the first value for the node (min = .75 x value and max = 1.5 x value) but it didn't give me the result expected.  With all the callbacks in the code, it is very tedious to trace the variables.  I'll dive back into it in a few days, but for now, I'm buried in tax preparation.  Keep my posted if you make progress.

**Comment by punitganshani** [comment]
@DanKline I tried fixing NaN issue in #238 

<img width="710" alt="spark-temp-humidity" src="https://cloud.githubusercontent.com/assets/1032553/13497463/4b9a8e64-e190-11e5-825f-ef96ddbcc456.PNG">

**Comment by DanKline** [comment]
@punitganshani I noticed you aren't displaying averages.  Does the chart resize when the scale is exceeded?    I wasn't able to get the averages to not render.  When I turned off the aggregator analytics job, the scales never rendered resulting in a plot that was never rendered.  I'm curious how you accomplished this.  I found the averages annoying at best and would rather not have them render.  I think they were added to sample just for the stream analytics demonstration.

**Comment by punitganshani** [comment]
@DanKline yes the scale changes. The `scaleMargin` was becoming `Infinity` so the domain had invalid range values. `self._y0` evaluation was giving `NaN` which was not rendering the values correctly. Yet to try the averages but this resolved the `NaN` issue.

**Comment by DanKline** [comment]
Thanks.  Is it possible to get a copy of what you have done?

**Comment by punitganshani** [comment]
Yes, the commit is merged into this repo (see #238) so you can try with the latest code. Average Stream Analytics works with this as well.

<img width="741" alt="average" src="https://cloud.githubusercontent.com/assets/1032553/13498732/234bf23a-e196-11e5-8467-5521729cea5c.PNG">

**Comment by DanKline** [comment]
Thanks ... I'll try it.

**Comment by DanKline** [comment]
Congratulations @punitganshani.  Nice job on the scale redraw issue.  I even held my finger on the sensor to cause the temp and humidity to climb rapidly, and the scaling and redraw worked perfectly.  I'm still seeing an issue with the averages, but curiously, not on the leading edge of the rendering, but rather on the redraw of previous data.  If time allows, I'll set up a stream capture and compare it against the D3 render data for the values.  I suspect that it may have to do with a mismatch between timing of the node data and the streaming analytic data sampling.  I'll let you know if I figure it out.  Thanks again.  That was an awesome piece of work.

**Comment by DanKline** [comment]
Here's a capture of the average rendering issue.

![image](https://cloud.githubusercontent.com/assets/3081226/13586563/4d262ae6-e487-11e5-87f1-a0e8e5c1dc91.png)

**Comment by DanKline** [comment]
Interestingly, if I leave the application rendering for a long enough time, the averages disappear from the renderings.  I'll have to capture the streams to see if we have another NaN issue with the Stream Analytics data.

![image](https://cloud.githubusercontent.com/assets/3081226/13588489/ae1393e4-e495-11e5-8019-86820c0dbbe0.png)

**Comment by DanKline** [comment]
I have stumbled, somewhat intuitively on the solution to the saw tooth averages in my renderings and would appreciate if anyone can verify my assumptions before I make recommendation for changes to the project.  In the configuration for the BI service I added some handling for late arrivals and it appears to have stabilized the renderings.  Can someone try this in their code and tell me if the result are consistent.
![ctdwdncg4pbi](https://cloud.githubusercontent.com/assets/3081226/14695838/74ca5eec-073b-11e6-9b50-2bfe2aabb3d8.png)
![image](https://cloud.githubusercontent.com/assets/3081226/14695855/86c8d498-073b-11e6-8c73-e7dfedc12234.png)
